substrate/frame/balances/src/impl_fungible.rs

```diff
@@ -174,7 +174,10 @@ impl<T: Config<I>, I: 'static> fungible::Unbalanced<T::AccountId> for Pallet<T,
 	}
 
 	fn deactivate(amount: Self::Balance) {
-		InactiveIssuance::<T, I>::mutate(|b| b.saturating_accrue(amount));
+		InactiveIssuance::<T, I>::mutate(|b| {
+			// InactiveIssuance cannot be greater than TotalIssuance.
+			*b = b.saturating_add(amount).min(TotalIssuance::<T, I>::get());
```

Comment:
do we have/had a problem with the current impl? why we not needs the same for `reactivate`?
if we have a wrong state now, I think it important to understand why, and address the cause. 
if no, I would just add some asserts on try-state, and leave these mutates as it is

Reply:
> do we have/had a problem with the current impl

Yes, it is a buggy implementation of the fungible trait because calling `deactivate` with an amt > than the current issuance will break things. See the tests referenced in the PR description for an example. 

> why we not needs the same for `reactivate`?

Because `reactivate` uses `saturating_reduce` to saturate `InactiveIssuance` at 0.

> if we have a wrong state now, I think it important to understand why, and address the cause.

This is the root cause of a buggy impl of fungible.

> if no, I would just add some asserts on try-state, and leave these mutates as it is

I disagree here. The idea behind these conformance tests is that we check that implementations implement the spec and behave as expected. If we left the mutates as-is, we effectively would need to allow this bug to exist in all implementations of fungibles. Let me know if I'm missing something though.

Reply:
I mean why would this function receive amount higher than `total_issuance`?
Are we sure it is fine to apply the min here? do clients expect such behaviour or at least ok with it?
Do we mention this in docs?
These questions I will ask myself before applying this.
Clients do not get any result from this function, with actual amount for example, they might not expect this.
If you double check all these, I am fine with this changes

Reply:
Let's step back and look at this from the POV of the fungible implementation. Without my fix, `deactivate` and `reactivate` work like this:

```rust
// Setting total issuance to 10 sets total issuance and active issuance to 10
T::set_total_issuance(10.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 10.into()); 

// Calling `deactivate` decreases active issuance by the amount passed in
T::deactivate(2.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 8.into()); 

// Calling `reactivate` saturates at `total_issuance`  
T::reactivate(4.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 10.into()); 

// Once saturated, calling `deactivate` reduces the `active_issuance` by the amount passed in  
T::deactivate(1.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 9.into()); 

// Calling `deactivate` with an amount greater than `active_issuance` makes `active_issuance` saturate at 0
// In reality, the internal implementation of active issuance in Balances pallet has created something like an underflow
T::deactivate(15.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 0.into()); 

// Calling `reactivate` does NOT increment `active_issuance` by the amount passed in, because the underlying value tracking it underflowed. 
// I see this as a bug, and an unacceptable implementation of fungible imo. 
T::reactivate(3.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 0.into()); 
```

So as you can see, currently activating over the `total_issuance` saturates correctly, but deactivating below `active_issuance` makes the state broken, where `reactivate` no longer works as expected.

This is an side-effect of how the Balances pallet implements the fungible trait and nothing else. I don't see how it can be anything other than a bug, and unacceptable to any fungible conformance tests. 

To pass the fungible conformance tests, imo this MUST hold:
```rust
assert_eq!(T::active_issuance(), 0.into());
T::reactivate(3.into()); 
assert_eq!(T::total_issuance(), 10.into()); 
assert_eq!(T::active_issuance(), 3.into());
```

which is what is achieved by the bug fix here.

Reply:
agree, let ensure we have this behaviour documented, that client can deactivate at most the total_issuance